### ISSUE TITLE
Prevent AST nodes from being modified by rules

### ISSUE THREAD

**Issue by lo1tuma** [eslint/eslint#1231]
In #1224 we discovered a bug where a rule modified an AST node which is passed to that rule. To prevent such bugs we could freeze the AST nodes or passing just a copy to the rules.

**Comment by gyandeeps** [comment]
In my case, because I did modified the AST (accidentally :smiley: ) it had adverse effects on the code following my code. That can be a rule or anything inside my rule.

**Comment by ilyavolodin** [comment]
I have a radical solution to that. How about we create our own plugin for ESLint that would check that node._anything_ never gets assigned?:-)

**Comment by michaelficarra** [comment]
Because

```js
function mutate(x) { x.y = 0; }
mutate(node);
```

**Comment by nzakas** [comment]
@gyandeeps see my previous comment re: freezing

**Comment by BYK** [comment]
>@gyandeeps see my previous comment re: freezing

Do you think we can get an option added to `espree` so it freezes the nodes as soon as it creates, thus avoiding the performance hit a bit may be?

**Comment by lo1tuma** [comment]
I don't think that would work because eslint modifies the AST nodes before
passing them to the rules.

**Comment by mysticatea** [comment]
I thought this problem should be checked by `eslint-tester`, and `eslint` does nothing in order to keep performance.
